Add Architecture Decision Record 001 - Pagemon integration with /status (#27)
bbc/status: docs/architecture/adr-001-pagemon.md

# Architecture Decision Record 001 - Pagemon integration with /status

## Context

> We would like to consume Pagemon data to surface Pagemon metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with a 'desktop' certificate. We want an approach that going forwards, will allow the /status Page to access other API's (data sources) e.g. CPS.

Reviewer: ```suggestion
We would like to consume Pagemon data to surface Pagemon metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with either a 'developer' certificate _or_ a 'desktop' certificate. We want an approach that going forwards, will allow the /status Page to access other API's (data sources) e.g. CPS.
```

Reviewer 2: > BBC staff with either a 'developer' certificate _or_ a 'desktop' certificate.

@benwainwright  Doesn't `/status` only request a desktop cert at the moment?

Reviewer: @abrarkhan1234 Quite right, my mistake.

> We would like to consume Pagemon data to surface Pagemon metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with a 'desktop' certificate. We want an approach that going forwards, will allow the /status Page to access other API's (data sources) e.g. CPS.

Reviewer: ```suggestion
We would like to consume [Pagemon](https://pagemon.tools.bbc.co.uk/) data to surface metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with a 'desktop' certificate. We want an approach that going forwards, will allow the /status Page to access other API's (data sources) e.g. CPS.
```

> We would like to consume [Pagemon](https://pagemon.tools.bbc.co.uk/) data to surface metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with a 'desktop' certificate. We want an approach that going forwards, will allow the /status Page to access other API's (data sources) e.g. CPS.

Reviewer: ```suggestion
We would like to consume Pagemon data to surface Pagemon metrics on the /status page. Pagemon requires a 'developer' certificate however we want to provide access to the /status page for any member of BBC staff with a 'desktop' certificate. We want an approach that, going forwards, will allow the /status page to access other APIs (data sources) e.g. CPS.```

There you go; there's a suggestion without the dev cert bit.

### Options

1. Proxy in front of the API

> A REST API service which would invoke the Pagemon JSON API with a developer certificate. This would allow a BBC staff member with only a desktop certificate, to make requests to the service, which would then in turn make requests to the Pagemon JSON API with a developer certificate. This would also support other internal BBC services e.g. CPS.

Reviewer: ```suggestion
A REST API service which would invoke the Pagemon JSON API with a developer certificate. This would allow a BBC staff member with only a desktop certificate to make requests to the service, which would then in turn make requests to the Pagemon JSON API with a developer certificate. This would also support other internal BBC services e.g. CPS.
```

2. AWS API Gateway with AWS Lambda

> A (store the data):

Reviewer: To the casual skimmer; the structure of this section isn't very clear. When I first read it, I thought I was reading a section with two alternate options, not components. I've suggested below a rewrite:

```suggestion
This approach is split into two components

* A service (Lambda) that is invoked every 60 seconds, which makes requests to Pagemon (and potentially other sources in future), fetches data and stores this as a JSON file in S3
* An API gateway endpoint that allows the `/status` endpoint to be consumed by the frontend React application

This approach decouples the request of the data from the retrieving of it, so will ensure speedy access regardless of how many data sources we integrate in the future. 
```

Reviewer 3: Agreed

Reviewer: What I've suggested above is basically what you've written in a different format. But thinking about it; we are using Broxy, not API gateway, aren't we?

Reviewer 2: As discussed below this approach was required if we needed to do any transform on the saved data, but now that we agree we will do any necessary transform when creating we can stick with accessing JSON through Broxy.

> Service (Lambda) that is invoked every 60 seconds, and makes requests to Pagemon (and in future further API sources), fetches data and stores this as a JSON file in S3.

Reviewer: ```suggestion
```

Reviewer 2: @benwainwright why are you suggesting this be removed?

> B (Consume the data):

Reviewer: ```suggestion
```

Reviewer 2: same question, why?

> The /status page would invoke an API Gateway end-point, which would invoke a Lambda which reads the JSON data from S3. The front-end /status page uses this JSON data to render the UI.

Reviewer: ```suggestion
```

Reviewer 2: @benwainwright Aah ok you have rewritten it here...

3. AWS Cognito

Supports user sign up, sign in and authentication with Oauth 2.0, SAML 2.0, and OpenID Connect. This would require custom work to integrate with BBC developer certificate.

### Decision

1. Proxy in front of the API

> Seems like too much effort initially. Would need to support a new service which has ongoing support required. We believe there is a simple approach available.

Reviewer: 'A simple approach' - doesn't seem to read right. Do you mean this?

```suggestion
Seems like too much effort initially. Would need to support a new service which has ongoing support required. We believe there is a simpler approach available.
```
